numpy/lib/function_base.py

```diff
@@ -3695,78 +3981,96 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
     if axis is None:
         axis = 0
 
-    Nx = ap.shape[axis]
-    indices = q * (Nx - 1)
+    if weights is None:
+        Nx = ap.shape[axis]
+        indices = q * (Nx - 1)
+    else:
+        # axis will be moved to -1 and indices will be floats
+        ap, indices, Nx = _indices_for_weighted_quantile(ap, q, axis, weights)
 
-    # round fractional indices according to interpolation method
     if interpolation == 'lower':
-        indices = floor(indices).astype(intp)
+        indices = np.floor(indices).astype(intp)
     elif interpolation == 'higher':
-        indices = ceil(indices).astype(intp)
+        indices = np.ceil(indices).astype(intp)
     elif interpolation == 'midpoint':
-        indices = 0.5 * (floor(indices) + ceil(indices))
+        indices = 0.5 * (np.floor(indices) + np.ceil(indices))
     elif interpolation == 'nearest':
-        indices = around(indices).astype(intp)
+        indices = np.around(indices).astype(intp)
     elif interpolation == 'linear':
         pass  # keep index as fraction and interpolate
     else:
         raise ValueError(
             "interpolation can only be 'linear', 'lower' 'higher', "
             "'midpoint', or 'nearest'")
 
-    n = np.array(False, dtype=bool) # check for nan's flag
+    n = np.array(False, dtype=bool)  # check for nan's flag
+    inexact = np.issubdtype(a.dtype, np.inexact)  # check if array has nan's
```

Comment:
This comment is misleading - you're checking if it _could_ have NaNs

Reply:
Sounds good.  Fixed.